From 7841ddb4589c084c929fce01a2d65a7c2b9f83db Mon Sep 17 00:00:00 2001
From: Ido Yariv <ido@wizery.com>
Date: Tue, 27 Sep 2011 14:20:27 +0300
Subject: [PATCH 09/18] board-4430: Add support for mmc5/dat1 irq source
 *PRELIMINARY*

Set dat1_gpio and implement a remuxing function for using mmc5/dat1 as an
irq source by omap_hsmmc

Signed-off-by: Ido Yariv <ido@wizery.com>
Signed-off-by: Eliad Peller <eliad@wizery.com>
---
 arch/arm/mach-omap2/board-4430sdp.c |   28 ++++++++++++++++++++++++++++
 1 files changed, 28 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/board-4430sdp.c b/arch/arm/mach-omap2/board-4430sdp.c
index 5b5d839..79012ca 100644
--- a/arch/arm/mach-omap2/board-4430sdp.c
+++ b/arch/arm/mach-omap2/board-4430sdp.c
@@ -94,6 +94,7 @@
 
 #define GPIO_WIFI_PMENA		54
 #define GPIO_WIFI_IRQ		53
+#define GPIO_MMC5_DAT1		148
 #define OMAP_HDMI_HPD_ADDR	0x4A100098
 #define OMAP_HDMI_PULLTYPE_MASK	0x00000010
 
@@ -432,6 +433,29 @@ static struct twl4030_usb_data omap4_usbphy_data = {
 	.phy_suspend	= omap4430_phy_suspend,
 };
 
+static void remux_mmc5_dat1(bool clocks_enabled)
+{
+	void __iomem *mux_base = NULL;
+	u32 val;
+
+	mux_base = ioremap(0x4A100000, 0x1000);
+	if (!mux_base) {
+		pr_err("%s: Could not remap mux base", __func__);
+		return;
+	}
+
+	val = __raw_readl(mux_base + 0x14C) & 0x0000FFFF;
+	if (clocks_enabled)
+		val |= (OMAP_MUX_MODE0 | OMAP_PIN_INPUT_PULLUP) << 16;
+	else
+		val |= (OMAP_MUX_MODE3 | OMAP_PIN_INPUT_PULLUP |
+			OMAP_PIN_OFF_WAKEUPENABLE) << 16;
+
+	__raw_writel(val, mux_base + 0x14C);
+
+	iounmap(mux_base);
+}
+
 static struct omap2_hsmmc_info mmc[] = {
 	{
 		.mmc		= 2,
@@ -439,6 +463,7 @@ static struct omap2_hsmmc_info mmc[] = {
 					MMC_CAP_1_8V_DDR,
 		.gpio_cd	= -EINVAL,
 		.gpio_wp	= -EINVAL,
+		.gpio_dat1	= -EINVAL,
 		.nonremovable   = true,
 		.ocr_mask	= MMC_VDD_29_30,
 		.no_off_init	= true,
@@ -448,6 +473,7 @@ static struct omap2_hsmmc_info mmc[] = {
 		.caps		= MMC_CAP_4_BIT_DATA | MMC_CAP_8_BIT_DATA |
 					MMC_CAP_1_8V_DDR,
 		.gpio_wp	= -EINVAL,
+		.gpio_dat1	= -EINVAL,
 	},
 	{
 		.mmc		= 5,
@@ -455,8 +481,10 @@ static struct omap2_hsmmc_info mmc[] = {
 				  MMC_CAP_SDIO_IRQ | MMC_CAP_ASYNC_SDIO_IRQ,
 		.gpio_cd	= -EINVAL,
 		.gpio_wp	= -EINVAL,
+		.gpio_dat1	= GPIO_MMC5_DAT1,
 		.ocr_mask	= MMC_VDD_165_195,
 		.nonremovable	= true,
+		.remux_dat1	= remux_mmc5_dat1,
 	},
 	{}	/* Terminator */
 };
-- 
1.7.6.401.g6a319

